sympy/core/numbers.py

```diff
@@ -438,6 +439,142 @@ def ilcm(*args):
     return a
 
 
+def igcdLLL(s, m1=2, n1=3, optimize=False):
+    """return g, B where g is the gcd of the integers in s and the
+    LLL-reduced basis and small multipliers such that those
+    multiples of the corresponding values of s is g.
+
+    If there are three values in s then the smallest multipliers
+    will be returned unless optimize is not True.
+
+    This is an iterative procedure whose quality in returning
+    small multipliers is related to alpha = m1/n1 which must
+    be in the range [3/8, 1]. When 1, the algorithm will iterate
+    longer but is not guaranteed to return the smallest multipliers
+    except for the case of 3 values in s.
+
+    Examples
+    ========
+
+    >>> from sympy.core.numbers import igcdLLL
+    >>> from sympy import Add, Matrix, symbols
+
+    >>> s = 2, 4, 6
+    >>> x = symbols('x:%s' % len(s))
+    >>> eq = Add(*[si*xi for si, xi in zip(s, x)]); eq
+    2*x0 + 4*x1 + 6*x2
+
+    >>> g, B = igcdLLL(s)
+    >>> assert Add(*[si*mi for si, mi in zip(s, B[-1])]) == g
```

Comment:
These examples seem almost designed to obfuscate what this function is doing.  At least a simple example that just directly shows the input and output would be a lot clearer:
```python
In [4]: gcd([2, 4, 6])
Out[4]: 2

In [5]: igcdLLL([2, 4, 6])
Out[5]: (2, [[1, 1, -1], [2, -1, 0], [1, 0, 0]])

In [6]: 1*2 + 1*4 + (-1)*6
Out[6]: 0
```
Using `Add` and other constructs complicates things.

Reply:
What exactly is this a basis for? Is `B[-1]` guaranteed to give the `Bezout` coefficients?

Reply:
There should be a reference to the paper for the algorithm

Reply:
It looks like `B` is the matrix sending the input numbers to `[0, ..., g].T`:
```python
In [5]: from sympy.core.numbers import igcdLLL

In [6]: nums = [3, 6, 12]

In [7]: g, B = igcdLLL(nums)

In [8]: g
Out[8]: 3

In [9]: B
Out[9]: [[-2, 1, 0], [0, -2, 1], [1, 0, 0]]

In [10]: Matrix(B) * Matrix(nums)
Out[10]: 
⎡0⎤
⎢ ⎥
⎢0⎥
⎢ ⎥
⎣3⎦
```
More generally I guess that if `nums` generates a Z-module then `B` is a map from that Z-module to the Z-module generated by `[0, ..., g]` where `g` is the gcd of nums.

Reply:
What about
```python
The following function will compute weighted sums of values:

>>> dot = lambda x, y: sum([i*j for i, j in zip(x, y)])
>>> dot((x, y), (2, 3))
2*x + 3*y

The greatest common divisor for elements of `s` and a list of multipliers are
computed by `igcdLLL`:

>>> g, B = igcdLLL(s)
>>> g
2
>>> B
[[1, 1, -1], [2, -1, 0], [1, 0, 0]]

The last row of `B` gives a list of small multipliers for the value of `s` that will produce a
sum equal to `g` while the others will produce a zero sum:

>>> [dot(s, b) for b in B]
[0, 0, 2]
```
Should it just stop there or shall I add
```python
One way these results can be used is to give a parametric solution for `2*x + 4*y + 6*z = c`.
Let `p0` and `p1` be the multiples for the first two rows; any mutliple of them will still sum to zero.
In order to get the sum to equal `c`, the last row of multipliers will be multiplied by `c/2`:

>>> gensol = (Matrix(B).T*Matrix([p0, p1, c/2])).T.tolist()[0]; gensol
[c/2 + p0 + 2*p1, p0 - p1, -p0]
>>> dot(s, _)
c
```

Reply:
Okay this is a bit clearer now. The first two rows of B are a basis for the integer solutions of `2*x + 4*y + 6*z = 0`. The last row of B gives coefficients for `gcd(2, 4, 6)` as a linear combination of 2, 4 and 6.

I'm not clear how exactly this is different from the Bezout coefficients:
https://en.wikipedia.org/wiki/B%C3%A9zout%27s_identity#For_three_or_more_integers

Reply:
>  how exactly this is different from the Bezout coefficients:

I don't know a lot about this, but I guess you could say the same thing about the extra parameters from the extended gcd calculation. For me they just appear to be roses called by other names.